### DISCUSSION TITLE
How to control the image after transform to show all original content before transform?

### DISCUSSION THREAD

**Discussion by nissansz** [python-pillow/Pillow#8103] [Q&A] [answered]
![23](https://github.com/python-pillow/Pillow/assets/41010669/8ab035f9-f0ee-4237-b4cc-2b1ee7aaea9f)

Sample code below may cause image after transform to show beyong original size.

```python
from PIL import Image
import numpy as np
import cv2

# 读取图像
image = Image.open(r"F:\23.png")

# 定义透视变换矩阵
matrix = np.array([
    [1, 0.2, 0],   # x 轴拉伸因子
    # [0.5, 1, 0],   # y 轴拉伸因子
    [0, 1, 0],   # y 轴拉伸因子
    [0, 0, 1]      # 平移因子
])


# # matrix = np.array([
# #     # [1, 0.5, 0],   # x 轴拉伸因子
# #     [0.5, 0, 0.5],   # x 轴拉伸因子
# #     # [0.5, 1, 0],   # y 轴拉伸因子
# #     [0, 1, 0],   # y 轴拉伸因子
# #     [0, 0, 1]      # 平移因子
# # ])
# 
# width, height = image.size
# # 计算变换后的图像尺寸
# new_width = int(width * 1.5)
# new_height = int(height * 1.5)
# # 进行透视变换
# # trans_image = image.transform(image.size, Image.AFFINE, matrix.flatten())
# trans_image = image.transform((new_width, new_height), Image.AFFINE, matrix.flatten())

# # 保存变换后的图像
# trans_image.save("output_image.jpg")
img = cv2.cvtColor(np.asarray(trans_image),cv2.COLOR_RGB2BGR)
cv2.imshow("OpenCV",img)
cv2.waitKey()
```

**Comment by radarhere** [comment, marked as answer]
```python
from PIL import Image
import numpy as np
image = Image.open("F23.png")
matrix = np.array([
  [1, 0.2, 0],
  [0, 1, 0],
  [0, 0, 1]
])
new_width = int(image.width * 1.5)
new_height = int(image.height * 1.5)
transformed_image = image.transform((new_width, new_height), Image.AFFINE, matrix.flatten())
transformed_image.save("output_image.jpg")
```
gives

![output_image](https://github.com/python-pillow/Pillow/assets/3112309/3b054059-6d20-4ba0-9b9a-645885d69f29)

You can use [`crop()`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.crop) to manually remove the black box.

But I think you're asking for a way to reverse the matrix operation? You can simply `flip()` the matrix.

```python
from PIL import Image
import numpy as np
image = Image.open("F23.png")
matrix = np.array([
  [1, 0.2, 0],
  [0, 1, 0],
  [0, 0, 1]
])
new_width = int(image.width * 1.5)
new_height = int(image.height * 1.5)
transformed_image = image.transform((new_width, new_height), Image.AFFINE, matrix.flatten())

matrix = np.flip(matrix)
reversed_image = image.transform(image.size, Image.AFFINE, matrix.flatten())
reversed_image.save("reversed_image.png")
```

**Reply by nissansz** [reply]
Your flip code does not have any transform effect.
I need transform effect like below, showing all text from original image.

![image](https://github.com/python-pillow/Pillow/assets/41010669/351932d6-d9a9-4a88-873f-15ee0ec0ec9d)

![image](https://github.com/python-pillow/Pillow/assets/41010669/2de9012c-2bbb-4875-afd2-301871fb79f3)
![image](https://github.com/python-pillow/Pillow/assets/41010669/cbd9fe05-e912-40f8-b5bd-85ff7e13c6d3)
![image](https://github.com/python-pillow/Pillow/assets/41010669/03e51841-0b4c-49c0-9b40-769409695bc9)
![image](https://github.com/python-pillow/Pillow/assets/41010669/0e51b5c2-5f8d-4abc-a69d-6513c82095f1)
![image](https://github.com/python-pillow/Pillow/assets/41010669/343e03ca-1d0d-42b1-a331-1fa4aceaa14d)
![image](https://github.com/python-pillow/Pillow/assets/41010669/7473e9fc-7c95-4f8f-9c82-35bc1caa4b35)

**Reply by radarhere** [reply]
When you asked
> How to control the image after transform to show all original content before transform?

I took this to mean that you wanted to apply a transformation, and then reverse that transformation to get the original image back again. My second code snippet does that, giving the original image again.

> I need transform effect like below, showing all text from original image.

I don't understand what you mean. My best guess is that you'd like to transform the background of the image, but not the text? As I have [said before](https://github.com/python-pillow/Pillow/issues/7376#issuecomment-1706317192), Pillow cannot detect text.

**Reply by nissansz** [reply]
I need to transform whole image including text, here is a wave effect from you before, I need more effect like above curved effect, etc.

![image](https://github.com/python-pillow/Pillow/assets/41010669/f91eb7d6-ca8a-4184-8aae-69f3f3535bab)

**Reply by nissansz** [reply]
There is a transform called perspective transformation, how to get such effect on whole image without losing text?

**Reply by radarhere** [reply]
If you would like a 'wave effect', I suggest you look at the `WaveDeformer` from https://github.com/python-pillow/Pillow/issues/7367#issuecomment-1706249174

> how to get such effect on whole image without losing text?

Again, Pillow cannot detect text.

**Reply by nissansz** [reply]
How to transform normal text to such effect?
![image](https://github.com/python-pillow/Pillow/assets/41010669/33a7e510-1b93-4ff1-915d-48cf8537d19c)

**Reply by radarhere** [reply]
Starting with [hopper.png](https://github.com/python-pillow/Pillow/blob/main/Tests/images/hopper.png)
```python
from PIL import Image, ImageTransform
im = Image.open("Tests/images/hopper.png")
transform = ImageTransform.QuadTransform((0, 0, 0, 128, 128, 158, 128, -30))
im.transform(im.size, transform).save("out.png")
```
gives
![out](https://github.com/python-pillow/Pillow/assets/3112309/cbdfa423-15ed-43a3-a5d2-8f22037717fb)

**Reply by nissansz** [reply]
Thank you, which parameters are fine to adjust QuadTransform((0, 0, 0, 128, 128, 158, 128, -30))?

**Reply by radarhere** [reply]
The numbers are x, y pairs of co-ordinates of the shape that is being transformed into the final image.

The original image is 128px by 128px.

The first two numbers are 0, 0, because the top left corner should stay the same.
The second two numbers are 0, 128, because the bottom left corner should stay the same.
The third two numbers are 128, 158, because I'm pulling in 30 pixels from below the image.
The fourth two numbers are 128, -30 because I'm pulling in 30 pixels from above the image.

**Reply by nissansz** [reply]
transform = ImageTransform.QuadTransform((0, 0, 0, height, width, height+30,  width, -30))
It works now. Thank you. 

![image](https://github.com/python-pillow/Pillow/assets/41010669/ca2d9219-e4c5-48aa-9af2-98c1edc411bc)



![23](https://github.com/python-pillow/Pillow/assets/41010669/5a622bc7-0232-45bc-bed1-e41f59e66893)

**Reply by nissansz** [reply]
> ```python
> from PIL import Image
> import numpy as np
> image = Image.open("F23.png")
> matrix = np.array([
>   [1, 0.2, 0],
>   [0, 1, 0],
>   [0, 0, 1]
> ])
> new_width = int(image.width * 1.5)
> new_height = int(image.height * 1.5)
> transformed_image = image.transform((new_width, new_height), Image.AFFINE, matrix.flatten())
> transformed_image.save("output_image.jpg")
> ```
> 
> gives
> 
> ![output_image](https://private-user-images.githubusercontent.com/3112309/336377493-3b054059-6d20-4ba0-9b9a-645885d69f29.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5OTgwNDAsIm5iZiI6MTcxNzk5Nzc0MCwicGF0aCI6Ii8zMTEyMzA5LzMzNjM3NzQ5My0zYjA1NDA1OS02ZDIwLTRiYTAtOWI5YS02NDU4ODVkNjlmMjkuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTBUMDUzNTQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTg5ZjE4ZjU4YzRkMGRhNDRkNGQ0YTNkZjA0ZTMyNjg2MGJjZjU0Njc0ZWE5NzBmNGY1NTZjMTgxNmVlYTEzZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.R4FbfB_XslMGj3Pmv6CyLBpFsjUJ1DkozSfzKo8_edY)
> 
> You can use [`crop()`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.crop) to manually remove the black box.
> 
> But I think you're asking for a way to reverse the matrix operation? You can simply `flip()` the matrix.
> 
> ```python
> from PIL import Image
> import numpy as np
> image = Image.open("F23.png")
> matrix = np.array([
>   [1, 0.2, 0],
>   [0, 1, 0],
>   [0, 0, 1]
> ])
> new_width = int(image.width * 1.5)
> new_height = int(image.height * 1.5)
> transformed_image = image.transform((new_width, new_height), Image.AFFINE, matrix.flatten())
> 
> matrix = np.flip(matrix)
> reversed_image = image.transform(image.size, Image.AFFINE, matrix.flatten())
> reversed_image.save("reversed_image.png")
> ```

matrix = np.array([
  [1, 0.2, 0],
  [0, 1, 0],
  [0, 0, 1]
])

If I change 0.2 to 0.8 etc., the image will be like below, some characters in the beiginning disappeared, how to solve such issue?
![image](https://github.com/python-pillow/Pillow/assets/41010669/c00fd43a-74fa-4b30-9216-ffca87640068)

**Reply by nissansz** [reply]
I copied original image to a white blank image width*1.2, then transform, can show whole image now
![image](https://github.com/python-pillow/Pillow/assets/41010669/124de429-b0ce-4566-aa67-0cce4063c232)

**Reply by radarhere** [reply]
Thank you for finding the solution.
